service/history/workflow_task_handler_callbacks.go

```diff
@@ -584,6 +585,18 @@ func (handler *workflowTaskHandlerCallbacksImpl) handleWorkflowTaskCompleted(
 			return nil, err
 		}
 
+		// Worker must respond with Update Accepted or Update Rejected message on every Update Requested
+		// message that were delivered on specific WT, when completing this WT.
+		// If worker ignored the update request (old SDK or SDK bug), then server fails the WT.
+		// Otherwise, this update will be delivered (and new WT created) again and again.
+		if err = workflowTaskHandler.ensureUpdatesProcessed(
+			ctx,
+			currentWorkflowTask.ScheduledEventID,
+			workflowTaskHeartbeating,
+		); err != nil {
+			return nil, err
+		}
```

Comment:
When I read this code and I do not apply internal knowledge of `ensureUpdatesProcessed` I think that the function tries to process all the unprocessed messages and, if it fails, returns an error. While the function does not try to process anything. It only checks whether there are unprocessed messages which need to be processed.
What I would do is to split the `ensure...` function into 2. One from the start to line 237 and call it `getUnprocessedIDs`. The rest I would call `builErrorMessage(unprocessedIDs)`.  Then here I would have:
```
if ids := workflowTaskHandler.getUnprocessedIDs(); len(ids) > 0 {
   return nil, workflowTaskHandler.buildErrorMessage(ids)
}
```
The same number of lines of code, but reads much better.

Reply:
As I changed the design, this method was also changed and it is much better now.